File: 2096-find-the-longest-valid-obstacle-course-at-each-position/2096-find-the-longest-valid-obstacle-course-at-each-position.py
class Solution:
    def longestObstacleCourseAtEachPosition(self, nums: List[int]) -> List[int]:
        dp = []
        ans = []

        for n in nums:
            if not dp or dp[-1] <= n:
                dp.append(n)
                ans.append(len(dp))
            else:
                idx = bisect_right(dp, n)
                ans.append(idx + 1)
                
                dp[idx] = n
        
        return ans

File: 2096-find-the-longest-valid-obstacle-course-at-each-position/2096-find-the-longest-valid-obstacle-course-at-each-position.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        vector<int> dp;
        vector<int> lengths;

        for (int i = 0; i < obstacles.size(); i++) {
            int lo = 0;
            int hi = dp.size() - 1;
            int ans = -1;
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (dp[mid] > obstacles[i]) {
                    hi = mid - 1;
                    ans = mid;
                } else {
                    lo = mid + 1;
                }
            }
            if (ans == -1) {
                dp.push_back(obstacles[i]);
                lengths.push_back(dp.size());
            } else {
                dp[ans] = obstacles[i];
                lengths.push_back(ans + 1);
            }
        }

        return lengths;
    }
};